File: alexandria/ui/src/io/intino/alexandria/ui/displays/components/List.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;
import io.intino.alexandria.ui.displays.Display;
import io.intino.alexandria.ui.displays.components.collection.Collection;
import io.intino.alexandria.ui.displays.components.collection.CollectionBehavior;
import io.intino.alexandria.ui.displays.events.collection.AddItemEvent;
import io.intino.alexandria.ui.displays.notifiers.ListNotifier;
import io.intino.alexandria.ui.model.Datasource;

public abstract class List<B extends Box, ItemComponent extends io.intino.alexandria.ui.displays.components.Item, Item> extends AbstractList<ListNotifier, B> implements Collection<ItemComponent, Item> {

    public List(B box) {
        super(box);
    }

    public List<B, ItemComponent, Item> source(Datasource source) {
        source(source, new CollectionBehavior<ItemComponent, Item>(this));
        return this;
    }

    @Override
    protected AddItemEvent itemEvent(Display display) {
        return new AddItemEvent(this, (ItemComponent)display, ((ItemComponent)display).item());
    }

}

File: framework/ui-elements/src/displays/components/List.js
import React from "react";
import {withStyles} from 'alexandria-ui-elements/src/util/muiStylesCompat';
import AutoSizer from 'react-virtualized-auto-sizer';
import AbstractList from "../../../gen/displays/components/AbstractList";
import ListNotifier from "../../../gen/displays/notifiers/ListNotifier";
import ListRequester from "../../../gen/displays/requesters/ListRequester";
import 'alexandria-ui-elements/res/styles/layout.css';
import {collectionPalette, CollectionStyles} from "./Collection";
import DisplayFactory from "alexandria-ui-elements/src/displays/DisplayFactory";

export const ListStyles = theme => ({
	...CollectionStyles(theme),
	itemView : {
		height: "100%",
		padding: "2px 12px",
		borderRadius: "14px",
		transition: "background-color 140ms ease, box-shadow 140ms ease",
		'&:hover' : {
			background: collectionPalette(theme).rowHoverBackground
		},
		'&:hover $selector' : {
			opacity: 1,
			pointerEvents: "auto"
		}
	},
});

export class EmbeddedList extends AbstractList {

	constructor(props) {
		super(props);
		this.notifier = new ListNotifier(this);
		this.requester = new ListRequester(this);
	};

	render() {
		const { classes } = this.props;
		return (<div ref={this.container} className={classes.collectionViewport + " flex"} style={{width:"100%",position:'relative',borderRadius:"18px"}}><AutoSizer>{({ height, width }) => (this.behavior.renderCollection(height, width))}</AutoSizer></div>);
	};

}

class List extends EmbeddedList {
    constructor(props) {
        super(props);
    }
}

export default withStyles(ListStyles, { withTheme: true })(List);
DisplayFactory.register("List", withStyles(ListStyles, { withTheme: true })(List));
